clip_benchmark/metrics/linear_probe.py

```diff
@@ -0,0 +1,211 @@
+    
+    probe = torch.nn.Linear(features[0].shape[0], targets.max().item() + 1)
+    devices = [x for x in range(torch.cuda.device_count())]
+    probe = probe.cuda()
+    probe = torch.nn.DataParallel(probe, device_ids=devices)
+    optimizer = torch.optim.AdamW(
+        probe.parameters(),
+        lr=lr,
+        weight_decay=0,
+    )
+    criterion = torch.nn.CrossEntropyLoss()
+
+    len_loader = len(feature_loader)
+    scheduler = cosine_lr(optimizer, lr, 0., epochs * len_loader)
+
+    for epoch in range(epochs):
+        end = time.time()
+        for i, (x, y) in enumerate(feature_loader):
+            x, y = x.cuda(), y.cuda()
+            step = i + epoch * len_loader
+            scheduler(step)
+            data_time = time.time() - end
+
+            optimizer.zero_grad()
+            with autocast():
+                pred = probe(x)
+                loss = criterion(pred, y)
+
+            loss.backward()
+            optimizer.step()
+
+            batch_time = time.time() - end
+            end = time.time()
+
+            if (i % 20) == 0:
+                num_samples = i * len(x)
+                samples_per_epoch = len(train_dataloader)
+                percent_complete = 100.0 * i / len(train_dataloader)
+                print(
+                    f"Train Epoch: {epoch} [{num_samples}/{samples_per_epoch} ({percent_complete:.0f}%)]\t"
+                    f"Loss: {loss.item():.6f}\tData (t) {data_time:.3f}\tBatch (t) {batch_time:.3f}\t"
+                    f"LR {optimizer.param_groups[0]['lr']:.5f}"
+                )
+
+    # finally, evaluate.
+    features = torch.load(os.path.join(feature_dir, 'features_val.pt'))
+    targets = torch.load(os.path.join(feature_dir, 'targets_val.pt'))
+    feature_dset = FeatureDataset(features, targets)
+    feature_loader = DataLoader(feature_dset, batch_size=batch_size, 
+                                    shuffle=True, num_workers=num_workers, 
+                                    pin_memory=True,
+                                )
+    true, pred = [], []
+    with torch.no_grad():
+        for x, y in tqdm(feature_loader):
+            x = x.to(device)
+            y = y.to(device)
+
+            with autocast():
+                # predict
+                logits = probe(x)
+
+            pred.append(logits.cpu())
+            true.append(y.cpu())
+            
+    logits = torch.cat(pred)
+    target = torch.cat(true)
+    pred = logits.argmax(axis=1)
+
+    # measure accuracy
+    if len(dataloader.dataset.classes) >= 5:
+        acc1, acc5 = accuracy(logits.float(), target.float(), topk=(1, 5))
+    else:
+        acc1, = accuracy(logits.float(), target.float(), topk=(1,))
+        acc5 = float("nan") 
+    mean_per_class_recall = balanced_accuracy_score(target, pred)
+    if verbose:
+        print(classification_report(target, pred, digits=3))
+
+    print('acc1:', acc1)
+    return {"acc1": acc1, "acc5": acc5, "mean_per_class_recall": mean_per_class_recall, 
```

Comment:
Might be best to pick different names for metrics here (like prefix with linear probe) so they don't collide with zero shot metrics when saving (see run.sh)